File: tests/15-basic-resolver.c
/**
 * Seccomp Library test program
 *
 * Copyright (c) 2012 Red Hat <pmoore@redhat.com>
 * Author: Paul Moore <paul@paul-moore.com>
 */

/*
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of version 2.1 of the GNU Lesser General Public License as
 * published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, see <http://www.gnu.org/licenses>.
 */

#include <errno.h>
#include <string.h>
#include <stdlib.h>

#include <seccomp.h>

unsigned int arch_list[] = {
	SCMP_ARCH_NATIVE,
	SCMP_ARCH_X86,
	SCMP_ARCH_X86_64,
	SCMP_ARCH_X32,
	SCMP_ARCH_ARM,
	SCMP_ARCH_AARCH64,
	SCMP_ARCH_MIPS,
	SCMP_ARCH_MIPS64,
	SCMP_ARCH_MIPS64N32,
	SCMP_ARCH_MIPSEL,
	SCMP_ARCH_MIPSEL64,
	SCMP_ARCH_MIPSEL64N32,
	SCMP_ARCH_PPC,
	SCMP_ARCH_PPC64,
	SCMP_ARCH_PPC64LE,
	SCMP_ARCH_S390,
	SCMP_ARCH_S390X,
	SCMP_ARCH_PARISC,
	SCMP_ARCH_PARISC64,
	SCMP_ARCH_RISCV64,
	-1
};

int main(int argc, char *argv[])
{
	int rc;
	int iter = 0;
	unsigned int arch;
	char *name = NULL;

	if (seccomp_syscall_resolve_name("open") != __SNR_open)
		goto fail;
	if (seccomp_syscall_resolve_name("read") != __SNR_read)
		goto fail;
	if (seccomp_syscall_resolve_name("INVALID") != __NR_SCMP_ERROR)
		goto fail;

	rc = seccomp_syscall_resolve_name_rewrite(SCMP_ARCH_NATIVE, "openat");
	if (rc != __SNR_openat)
		goto fail;

	while ((arch = arch_list[iter++]) != -1) {
		int sys;
		int nr_open;
		int nr_read;
		int nr_socket;
		int nr_shmctl;

		if (seccomp_syscall_resolve_name_arch(arch,
						      "INVALID") != __NR_SCMP_ERROR)
			goto fail;
		name = seccomp_syscall_resolve_num_arch(arch, __NR_SCMP_ERROR);
		if (name != NULL)
			goto fail;

		nr_open = seccomp_syscall_resolve_name_arch(arch, "open");
		if (nr_open == __NR_SCMP_ERROR)
			goto fail;
		nr_read = seccomp_syscall_resolve_name_arch(arch, "read");
		if (nr_read == __NR_SCMP_ERROR)
			goto fail;
		nr_socket = seccomp_syscall_resolve_name_rewrite(arch, "socket");
		if (nr_socket == __NR_SCMP_ERROR)
			goto fail;
		nr_shmctl = seccomp_syscall_resolve_name_rewrite(arch, "shmctl");
		if (nr_shmctl == __NR_SCMP_ERROR)
			goto fail;

		name = seccomp_syscall_resolve_num_arch(arch, nr_open);
		if (name == NULL || strcmp(name, "open") != 0)
			goto fail;
		free(name);
		name = NULL;

		name = seccomp_syscall_resolve_num_arch(arch, nr_read);
		if (name == NULL || strcmp(name, "read") != 0)
			goto fail;
		free(name);
		name = NULL;

		name = seccomp_syscall_resolve_num_arch(arch, nr_socket);
		if (name == NULL ||
		    (strcmp(name, "socket") != 0 &&
		     strcmp(name, "socketcall") != 0))
			goto fail;
		free(name);
		name = NULL;

		name = seccomp_syscall_resolve_num_arch(arch, nr_shmctl);
		if (name == NULL ||
		    (strcmp(name, "shmctl") != 0 && strcmp(name, "ipc") != 0))
			goto fail;
		free(name);
		name = NULL;

		/* socket pseudo-syscalls */
		if (seccomp_syscall_resolve_name_arch(arch, "socketcall") > 0) {
			for (sys = -101; sys >= -120; sys--) {
				name = seccomp_syscall_resolve_num_arch(arch,
									sys);
				if (name == NULL)
					goto fail;
				free(name);
				name = NULL;
			}
		}
		/* ipc pseudo-syscalls */
		if (seccomp_syscall_resolve_name_arch(arch, "ipc") > 0) {
			for (sys = -201; sys >= -204; sys--) {
				name = seccomp_syscall_resolve_num_arch(arch,
									sys);
				if (name == NULL)
					goto fail;
				free(name);
				name = NULL;
			}
			for (sys = -211; sys >= -214; sys--) {
				name = seccomp_syscall_resolve_num_arch(arch,
									sys);
				if (name == NULL)
					goto fail;
				free(name);
				name = NULL;
			}
			for (sys = -221; sys >= -224; sys--) {
				name = seccomp_syscall_resolve_num_arch(arch,
									sys);
				if (name == NULL)
					goto fail;
				free(name);
				name = NULL;
			}
		}
	}

	return 0;

fail:
	if (name != NULL)
		free(name);
	return 1;
}

File: tests/rust/15-basic-resolver.rs
// SPDX-License-Identifier: LGPL-2.1-only
//
// Copyright 2024 Sony Group Corporation
//
// Seccomp Library test program
//

use anyhow::Result;
use libseccomp::*;

const ARCH_LIST: [ScmpArch; 23] = [
    ScmpArch::Native,
    ScmpArch::X86,
    ScmpArch::X8664,
    ScmpArch::X32,
    ScmpArch::Arm,
    ScmpArch::Aarch64,
    ScmpArch::Loongarch64,
    ScmpArch::M68k,
    ScmpArch::Mips,
    ScmpArch::Mips64,
    ScmpArch::Mips64N32,
    ScmpArch::Mipsel,
    ScmpArch::Mipsel64,
    ScmpArch::Mipsel64N32,
    ScmpArch::Ppc,
    ScmpArch::Ppc64,
    ScmpArch::Ppc64Le,
    ScmpArch::S390,
    ScmpArch::S390X,
    ScmpArch::Parisc,
    ScmpArch::Parisc64,
    ScmpArch::Riscv64,
    ScmpArch::Sh,
];

fn main() -> Result<()> {
    assert_eq!(ScmpSyscall::from_name("open")?, ScmpSyscall::new("open"));
    assert_eq!(ScmpSyscall::from_name("read")?, ScmpSyscall::new("read"));
    assert!(ScmpSyscall::from_name("INVALID").is_err());
    assert_eq!(
        ScmpSyscall::from_name_by_arch_rewrite("openat", ScmpArch::Native)?,
        ScmpSyscall::new("openat")
    );

    for arch in ARCH_LIST {
        assert!(ScmpSyscall::from_name("INVALID").is_err());
        assert!(ScmpSyscall::from(-1).get_name_by_arch(arch).is_err());

        let nr_open = ScmpSyscall::from_name_by_arch("open", arch)?;
        let nr_read = ScmpSyscall::from_name_by_arch("read", arch)?;
        let nr_socket = ScmpSyscall::from_name_by_arch_rewrite("socket", arch)?;
        let nr_shmctl = ScmpSyscall::from_name_by_arch_rewrite("shmctl", arch)?;

        let mut name = nr_open.get_name_by_arch(arch)?;
        assert_eq!(name, "open");
        name = nr_read.get_name_by_arch(arch)?;
        assert_eq!(name, "read");
        name = nr_socket.get_name_by_arch(arch)?;
        assert!(name == "socket" || name == "socketcall");
        name = nr_shmctl.get_name_by_arch(arch)?;
        assert!(name == "shmctl" || name == "ipc");

        // socket pseudo-syscalls
        if ScmpSyscall::from_name_by_arch("socketcall", arch)? > 0.into() {
            for sys in -120..-100 {
                ScmpSyscall::from(sys).get_name_by_arch(arch)?;
            }
        }

        // ipc pseudo-syscalls
        if ScmpSyscall::from_name_by_arch("ipc", arch)? > 0.into() {
            for sys in -204..-200 {
                ScmpSyscall::from(sys).get_name_by_arch(arch)?;
            }
            for sys in -214..-210 {
                ScmpSyscall::from(sys).get_name_by_arch(arch)?;
            }
            for sys in -224..-220 {
                ScmpSyscall::from(sys).get_name_by_arch(arch)?;
            }
        }
    }
    Ok(())
}
